src/py_zfs_pool.c

```diff
@@ -28,378 +24,257 @@ int py_zfs_pool_init(PyObject *type, PyObject *args, PyObject *kwds) {
+PyObject *py_zfs_pool_get_name(py_zfs_pool_t *self, void *extra) {
+	if (self == NULL)
+		Py_RETURN_NONE;
+	return Py_NewRef(self->name);
 }
 
 static
-PyObject *py_zfs_pool_sync(PyObject *self, PyObject *args) {
+PyObject *py_zfs_pool_asdict(PyObject *self, PyObject *args) {
 	Py_RETURN_NONE;
 }
 
 static
-PyObject *py_zfs_pool_create(PyObject *self, PyObject *args) {
-	Py_RETURN_NONE;
-}
+PyObject *py_zfs_pool_root_dataset(PyObject *self, PyObject *args) {
+	py_zfs_dataset_t *out = NULL;
+	py_zfs_pool_t *p = (py_zfs_pool_t *)self;
+	PyObject *fargs = PyTuple_New(0);
+	PyObject *fkwargs = PyDict_New();
 
-static
-PyObject *py_zfs_pool_attach_vdevs(PyObject *self, PyObject *args) {
-	Py_RETURN_NONE;
-}
+	fargs = PyTuple_New(0);
+	if (fargs != 0)
+		return (NULL);
 
-static
-PyObject *py_zfs_pool_sync_vdev_by_guid(PyObject *self, PyObject *args) {
-	Py_RETURN_NONE;
-}
+	fkwargs = PyDict_New();
+	if (fkwargs != 0) {
+		Py_DECREF(fargs);
+		return (NULL);
+	}
 
-static
-PyObject *py_zfs_pool_delete(PyObject *self, PyObject *args) {
-	Py_RETURN_NONE;
-}
+	if (PyDict_SetItemString(fkwargs, "name", p->name) != 0) {
+		Py_DECREF(fargs);
+		Py_DECREF(fkwargs);
+		return (NULL);
+	}
 
-static
-PyObject *py_zfs_pool_start_scrub(PyObject *self, PyObject *args) {
-	Py_RETURN_NONE;
-}
+	out = (py_zfs_dataset_t *)py_zfs_resource_open((PyObject *)p->pylibzfsp,
+	    fargs, fkwargs);
 
-static
-PyObject *py_zfs_pool_stop_scrub(PyObject *self, PyObject *args) {
-	Py_RETURN_NONE;
+	Py_DECREF(fargs);
+	Py_DECREF(fkwargs);
+	if (out == NULL)
+		return (NULL);
+	return ((PyObject *)out);
 }
 
 static
 PyObject *py_zfs_pool_clear(PyObject *self, PyObject *args) {
-	Py_RETURN_NONE;
-}
-
-static
-PyObject *py_zfs_pool_upgrade(PyObject *self, PyObject *args) {
-	Py_RETURN_NONE;
-}
-
-static
-PyObject *py_zfs_pool_ddt_prefetch(PyObject *self, PyObject *args) {
-	Py_RETURN_NONE;
-}
-
-static
-PyObject *py_zfs_pool_ddt_prune(PyObject *self, PyObject *args) {
-	Py_RETURN_NONE;
-}
-
-static
-PyObject *py_zfs_pool_get_root_dataset(py_zfs_prop_t *self, void *extra) {
-	Py_RETURN_NONE;
-}
-
-static
-PyObject *py_zfs_pool_get_root_vdev(py_zfs_prop_t *self, void *extra) {
-	Py_RETURN_NONE;
-}
-
-static
-PyObject *py_zfs_pool_get_data_vdevs(py_zfs_prop_t *self, void *extra) {
-	Py_RETURN_NONE;
-}
-
-static
-PyObject *py_zfs_pool_get_log_vdevs(py_zfs_prop_t *self, void *extra) {
-	Py_RETURN_NONE;
-}
-
-static
-PyObject *py_zfs_pool_get_cache_vdevs(py_zfs_prop_t *self, void *extra) {
-	Py_RETURN_NONE;
-}
-
-static
-PyObject *py_zfs_pool_get_spare_vdevs(py_zfs_prop_t *self, void *extra) {
-	Py_RETURN_NONE;
-}
-
-static
-PyObject *py_zfs_pool_get_special_vdevs(py_zfs_prop_t *self, void *extra) {
-	Py_RETURN_NONE;
-}
-
-static
-PyObject *py_zfs_pool_get_dedup_vdevs(py_zfs_prop_t *self, void *extra) {
-	Py_RETURN_NONE;
-}
-
-static
-PyObject *py_zfs_pool_get_groups(py_zfs_prop_t *self, void *extra) {
-	Py_RETURN_NONE;
-}
+	int ret = 0;
+	nvlist_t *policy = fnvlist_alloc();
```

Comment:
nvlist related operations should probably be done within `Py_BEGIN_ALLOW_THREADS` (where we drop GIL).

Reply:
Updated, can you please take a look?